src/rust/engine/rule_graph/src/rules.rs

```diff
@@ -22,26 +23,61 @@ pub trait TypeId:
     I: Iterator<Item = Self>;
 }
 
-pub trait DependencyKey:
-  Clone + Copy + Debug + Display + Hash + Ord + Eq + Sized + 'static
-{
-  type TypeId: TypeId;
+#[derive(DeepSizeOf, Eq, Hash, PartialEq, Clone, Debug, PartialOrd, Ord)]
+pub struct DependencyKey<T: TypeId> {
+  product: T,
+  provided_params: SmallVec<[T; 2]>,
+}
 
-  ///
-  /// Generate a DependencyKey for a dependency at the "root" of the RuleGraph, which represents an
-  /// entrypoint into the set of installed Rules.
-  ///
-  fn new_root(product: Self::TypeId) -> Self;
+impl<T: TypeId> DependencyKey<T> {
+  pub fn new(product: T) -> Self {
+    DependencyKey {
+      product,
+      provided_params: SmallVec::default(),
+    }
+  }
+
+  pub fn new_with_params<I: IntoIterator<Item = T>>(product: T, provided_params: I) -> Self {
+    let mut provided_params = provided_params.into_iter().collect::<SmallVec<[T; 2]>>();
+    provided_params.sort();
+
+    if cfg!(debug_assertions) {
```

Comment:
This could also be written without the `if` as: 

```
#[cfg(debug_assertions)]
{
 ...
}
```

Rust will probably figure out to elide the block given the condition being known at compile time, but this latter form more directly represents conditional compilation to me as a reader (just like using `#[cfg(test)]` in other cases).